pkg/tfshim/sdk-v2/upgrade_state.go

```diff
@@ -6,92 +6,149 @@ import (
+}
 
-	// Strip out removed fields.
-	schema.RemoveAttributes(ctx, json, configBlock.ImpliedType())
+func extractSchemaVersion(meta map[string]any) (int64, bool, error) {
+	versionValue, ok := meta["schema_version"]
+	if !ok {
+		return 0, false, nil
+	}
 
-	// now we need to turn the state into the default json representation, so
-	// that it can be re-decoded using the actual schema.
-	v, err := schema.JSONMapToStateValue(json, configBlock)
+	versionString, ok := versionValue.(string)
+	if !ok {
+		return 0, true, fmt.Errorf("unexpected type %T for schema_version", versionValue)
+	}
+	v, err := strconv.ParseInt(versionString, 0, 32)
 	if err != nil {
-		return nil, err
+		return 0, true, err
+	}
+	return v, true, nil
+}
+
+func upgradeResourceState(ctx context.Context, t string, p *schema.Provider, res *schema.Resource,
+	instanceState *terraform.InstanceState) (*terraform.InstanceState, error) {
+
+	if instanceState == nil {
+		return nil, nil
 	}
 
-	// Now we need to make sure blocks are represented correctly, which means
-	// that missing blocks are empty collections, rather than null.
-	// First we need to CoerceValue to ensure that all object types match.
-	v, err = configBlock.CoerceValue(v)
+	rawState := instanceState.RawState
+
+	// If RawState is not set but attributes is, we need to hydrate RawState
+	// from attributes.
+	if rawState.IsNull() {
+		// We default to assuming that the old state has the same shape as the new
+		// resource.
+		typ := res.CoreConfigSchema().ImpliedType()
+
+		// Find the version, if present, to deserialize instanceState into.
+		version, hasVersion, err := extractSchemaVersion(instanceState.Meta)
+		if err != nil {
+			return nil, err
+		}
+
+		// If we have a version, we use the schema shape that matches the version
+		// specified.
+		if hasVersion {
+			for _, t := range res.StateUpgraders {
+				if t.Version == int(version) {
+					typ = t.Type
+					break
+				}
+			}
+		}
+
+		rawState, err = instanceState.AttrsAsObjectValue(typ)
+		if err != nil {
+			return nil, fmt.Errorf("state from attributes: %w", err)
+		}
+	}
+
+	if state := rawState.AsValueMap(); !has(state, "id") {
+		state["id"] = cty.StringVal(instanceState.ID)
+		rawState = cty.ObjectVal(state)
+	}
+
+	v, newMeta, err := upgradeResourceStateGRPC(ctx, t, res, rawState, instanceState.Meta, p.GRPCProvider())
 	if err != nil {
 		return nil, err
 	}
 
-	// Normalize the value and fill in any missing blocks.
-	v = schema.NormalizeObjectFromLegacySDK(v, configBlock)
-
 	// Convert the value back to an InstanceState.
 	newState, err := res.ShimInstanceStateFromValue(v)
 	if err != nil {
 		return nil, err
 	}
-	newState.RawConfig = instanceState.RawConfig
 
-	// Copy the original ID and meta to the new state and stamp in the new version.
+	newState.RawConfig = instanceState.RawConfig
+	newState.RawState = v
+	newState.Meta = newMeta
 	newState.ID = instanceState.ID
 
 	// If state upgraders have modified the ID, respect the modification.
```

Comment:
I vaguely recall working on the fix for this, would be interesting to see if the gRPC version just handles it right.

Reply:
It doesn't, since the gRPC version returns a `(cty.Value, map[string]any, error)`. This is still needed to bridge back to `*terraform.InstanceState`.

Reply:
Brilliant. We can connect this to https://github.com/pulumi/pulumi-terraform-bridge/issues/1895 - I believe if we do this right and transition to PlanResourceChange cty.Value is all that is needed.